src/main/java/org/deepsymmetry/beatlink/AnnouncementSocketConnection.java

```diff
@@ -0,0 +1,294 @@
+                                stop();
+                            }
+                            received = false;
+                        }
+                        try {
+                            if (received) {
+                                final Util.PacketType kind = Util.validateHeader(packet, ANNOUNCEMENT_PORT);
+                                if (kind == Util.PacketType.DEVICE_KEEP_ALIVE) {
+                                    // Looks like the kind of packet we need
+                                    if (packet.getLength() < 54) {
+                                        logger.warn("Ignoring too-short " + kind.name + " packet; expected 54 bytes, but only got " +
+                                                packet.getLength() + ".");
+                                    } else {
+                                        if (packet.getLength() > 54) {
+                                            logger.warn("Processing too-long " + kind.name + " packet; expected 54 bytes, but got " +
+                                                    packet.getLength() + ".");
+                                        }
+                                        DeviceAnnouncement announcement = new DeviceAnnouncement(packet);
+
+                                        deliverDeviceAnnouncement(announcement);
+                                    }
+                                } else if (kind == Util.PacketType.DEVICE_HELLO) {
+                                    logger.debug("Received device hello packet.");
+                                } else if (kind != null) {
+                                    // TODO, probably should be a set of listeners to not couple VirtualCdj to this class
+                                    VirtualCdj.getInstance().handleSpecialAnnouncementPacket(kind, packet);
+                                }
+                            }
+                        } catch (Throwable t) {
+                            logger.warn("Problem processing DeviceAnnouncement packet", t);
+                        }
+                    }
+                }
+            }, "beat-link DeviceFinder receiver");
+            receiver.setDaemon(true);
+            receiver.start();
+        }
+    }
+
+    public void send(DatagramPacket packet) throws IOException {
+        announcementSocket.get().send(packet);
+    }
+
+    @Override
+    public SocketSender getSocketSender() {
+        return this;
+    }
+
+
+    /**
+     * Keeps track of the registered device announcement listeners.
+     */
+    private final Set<AnnouncementListener> deviceListeners =
+            Collections.newSetFromMap(new ConcurrentHashMap<AnnouncementListener, Boolean>());
+
+    /**
+     * Adds the specified device announcement listener to receive device announcements when DJ Link devices
+     * are found on or leave the network. If {@code listener} is {@code null} or already present in the list
```

Comment:
This is describing the `DeviceAnnouncementListener` interface offered by the `DeviceFinder`, and should be simplified here to say it just lets you know when packets are received. Also, how do you feel about calling this lower-level interface `AnnouncementPacketListener`?

Reply:
Love it, done.